internal/conpty/conpty.go

```diff
@@ -0,0 +1,153 @@
+	}
+
+	outPipeOurs, ptyOut, err := os.Pipe()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create pipes for pseudo console: %w", err)
+	}
+
+	var hpc windows.Handle
+	coord := windows.Coord{X: columns, Y: rows}
+	err = winapi.CreatePseudoConsole(coord, windows.Handle(ptyIn.Fd()), windows.Handle(ptyOut.Fd()), 0, &hpc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create pseudo console: %w", err)
+	}
+
+	// The pty's end of its pipes can be closed here without worry. They're duped into the conhost
+	// that will be launched and will be released on a call to ClosePseudoConsole() (Close() on the CPty object).
+	if err := ptyOut.Close(); err != nil {
+		return nil, fmt.Errorf("failed to close pseudo console handle: %w", err)
+	}
+	if err := ptyIn.Close(); err != nil {
+		return nil, fmt.Errorf("failed to close pseudo console handle: %w", err)
+	}
+
+	return &CPty{
+		hpc:     hpc,
+		inPipe:  inPipeOurs,
+		outPipe: outPipeOurs,
+	}, nil
+}
+
+// UpdateProcThreadAttribute updates the passed in attribute list to contain the entry necessary for use with
+// CreateProcess.
+func (c *CPty) UpdateProcThreadAttribute(attributeList *winapi.ProcThreadAttributeList) error {
+	c.handleLock.RLock()
+	defer c.handleLock.RUnlock()
+
+	if c.hpc == 0 {
+		return errClosedConPty
+	}
+
+	err := winapi.UpdateProcThreadAttribute(
+		attributeList,
+		0,
+		winapi.PROC_THREAD_ATTRIBUTE_PSEUDOCONSOLE,
+		unsafe.Pointer(c.hpc),
+		unsafe.Sizeof(c.hpc),
+		nil,
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to update proc thread attributes for pseudo console: %w", err)
+	}
+	return nil
+}
+
+// Resize resizes the internal buffers of the pseudo console to the passed in size
+func (c *CPty) Resize(columns, rows int16) error {
+	c.handleLock.RLock()
+	defer c.handleLock.RUnlock()
+
+	if c.hpc == 0 {
+		return errClosedConPty
+	}
+
+	coord := windows.Coord{X: columns, Y: rows}
+	if err := winapi.ResizePseudoConsole(c.hpc, coord); err != nil {
+		return fmt.Errorf("failed to resize pseudo console: %w", err)
+	}
+	return nil
+}
+
+// Close closes the pseudo-terminal and cleans up all attached resources
+func (c *CPty) Close() error {
+	c.handleLock.Lock()
+	defer c.handleLock.Unlock()
+
+	if c.hpc == 0 {
+		return errClosedConPty
+	}
+
+	// Close the pseudo console, set the handle to 0 to invalidate this object and then close the side of the pipes that we own.
+	winapi.ClosePseudoConsole(c.hpc)
+	c.hpc = 0
+	if err := c.inPipe.Close(); err != nil {
+		return fmt.Errorf("failed to close pseudo console input pipe: %w", err)
+	}
+	if err := c.outPipe.Close(); err != nil {
+		return fmt.Errorf("failed to close pseudo console output pipe: %w", err)
+	}
+	return nil
+}
+
+// OutPipe returns the output pipe of the pseudo console.
+func (c *CPty) OutPipe() *os.File {
+	return c.outPipe
+}
+
+// InPipe returns the input pipe of the pseudo console.
+func (c *CPty) InPipe() *os.File {
+	return c.inPipe
+}
```

Comment:
What will these be used for?

Reply:
To communicate with the pseudo console. Can think of OutPipe as stdout/err over one pipe, and InPipe as writing to stdin.